backend/shared/shared/oauth2/auth.py

```diff
@@ -0,0 +1,129 @@
+import logging
+
+import requests
+from django.contrib.auth.models import Group
+from django.core.exceptions import PermissionDenied
+from django_auth_adfs.backend import AdfsAuthCodeBackend
+from django_auth_adfs.config import provider_config, settings
+from django_auth_adfs.exceptions import MFARequired
+
+LOGGER = logging.getLogger(__name__)
+
+
+class HelsinkiAdfsAuthCodeBackend(AdfsAuthCodeBackend):
+    def get_member_objects_from_graph_api(
+        self, user_oid: str, graph_api_access_token: str
+    ):
+        url = f"https://graph.microsoft.com/v1.0/users/{user_oid}/getMemberObjects"
+
+        headers = {
+            "Authorization": f"Bearer {graph_api_access_token}",
+        }
+
+        data = {"securityEnabledOnly": False}
+
+        response = requests.post(
+            url,
+            json=data,
+            headers=headers,
+        )
+        response.raise_for_status()
+
+        # This will be a list of UUIDs. We will use them as group names in Django.
+        groups = response.json()["value"]
+        return groups
+
+    def update_user_groups_from_graph_api(
+        self, user, user_oid: str, graph_api_access_token: str
+    ):
+        groups = self.get_member_objects_from_graph_api(
+            user_oid, graph_api_access_token
+        )
+
+        django_groups = [group.name for group in user.groups.all()]
+
+        if sorted(groups) != sorted(django_groups):
+            existing_groups = list(Group.objects.filter(name__in=groups).iterator())
+            existing_group_names = frozenset(group.name for group in existing_groups)
+            new_groups = [
+                Group.objects.get_or_create(name=name)[0]
+                for name in groups
+                if name not in existing_group_names
+            ]
```

Comment:
The code doesn't ever seem to remove groups, should it remove groups if groups have been removed from the user in AD?

Reply:
Yep, I copied this from the library but seems it was not thought out there as well :grin: I'll add group removes as well.

Reply:
Fixed